core/tests/protractor_desktop/navigation.js

```diff
@@ -146,7 +146,39 @@ describe('Static Pages Tour', function() {
 
       await users.createAndLoginUser('user@navigation.com', 'navigationUser');
       await browser.get('/login');
+
+      // Login page performs client side redirection which is known to cause
+      // "both angularJS testability and angular testability are undefined"
+      // flake.
+      // As suggested by protractor devs here (http://git.io/v4gXM), waiting for
+      // angular is disabled during client side redirects.
+      await browser.waitForAngularEnabled(false);
+
+      // Waiting for first redirection (login page to splash page).
+      await browser.driver.wait(async() => {
```

Comment:
Here and elsewhere ... is it possible to use a "wait for thing to happen" construct? If you do it with a hard time limit, that can introduce flakiness due to slower machines (it's happened in the past).

DItto elsewhere.

Reply:
The "wait for thing to happen" construct was the first approach.
The issue with the first approach was that the protractor was waiting for events in the current page itself (not the new page after redirection). This introduced "document unloaded while waiting for result" flake. There is a race condition after the button is clicked between the protractor and angular code. When angular code wins we get no error (redirection done before protractor starts waiting) and error when protractor wins (protractor starts waiting before redirection is done).

Reply:
I found this waiting for redirection approach in the protractor codebase and is written by the protractor devs themself.
https://github.com/angular/protractor/blob/4bc80d1a459542d883ea9200e4e1f48d265d0fda/spec/withLoginConf.js#L29